forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt

```diff
@@ -32,68 +31,78 @@ internal object LDContextKind {
-    private var internalVaultType: VaultType? = null
+internal val ALWAYS_BT = 100.0
+internal val ALWAYS_VGS = 0.0
+
+internal fun computeVaultType(trafficPrimaryPercentFlag: Double): VaultType {
+    val randomNum = Math.random() * 100
+    return if (randomNum < trafficPrimaryPercentFlag) VaultType.BT_VAULT_TYPE else VaultType.VGS_VAULT_TYPE
+}
 
+internal object LDManager {
     private var internalLogger: Log = Log.getSilentInstance()
     private var client: LDClient? = null
 
-    internal var vaultType: VaultType?
-        get() = internalVaultType
-
-        // The setter is only exposed for testing purposes. Otherwise, it is entirely internal and
-        // shouldn't be used directly.
-        set(value) {
-            if (value == null) {
-                internalLogger.w("[LaunchDarkly] vaultType is being reset to null. This should only happen while unit testing!")
-                internalVaultType = null
-            } else if (internalVaultType == null) {
-                internalVaultType = value
-            } else {
-                throw Error("vaultType can only be set once!")
-            }
-        }
+    // default to 100% VGS usage in case LD flag retrieval fails
+    private var primaryTrafficPercent = ALWAYS_VGS
 
-    internal fun initialize(app: Application, logger: Log, dataSource: TestData? = null) {
-        val ldConfig = if (dataSource != null) {
-            LDConfig.Builder()
-                .mobileKey(LD_MOBILE_KEY)
-                .dataSource(dataSource)
-                .build()
-        } else {
-            LDConfig.Builder()
-                .mobileKey(LD_MOBILE_KEY)
-                .build()
-        }
+    // we only ever want to fetch primary traffic percent flag once
+    // so that all parts of the code are guaranteed to use the same
+    // value
+    private var hasFetchedPrimaryTrafficPercentFlag = false
+
+    fun createLdConfig(ldMobileKey: String): LDConfig {
+        return LDConfig.Builder()
+            .mobileKey(ldMobileKey)
+            .build()
+    }
+    fun TEST_createLdConfig(ldMobileKey: String, dataSource: TestData): LDConfig {
+        return LDConfig.Builder()
+            .mobileKey(ldMobileKey)
+            .dataSource(dataSource)
+            .build()
+    }
+
+    internal fun initialize(app: Application, logger: Log, ldConfig: LDConfig) {
         val contextKind = ContextKind.of(LDContextKind.SERVICE)
         val context = LDContext.create(contextKind, LDContexts.ANDROID_CONTEXT)
         client = LDClient.init(app, ldConfig, context, 0)
 
         internalLogger = logger
     }
 
-    // vaultType is instantiated lazily and is a singleton. Once we set the vault type once, we don't
-    // want to overwrite it! We must take in the application as a parameter, which means that a
-    // ForagePINEditText must be rendered before any of the ForageSDKApi functions are called.
+    internal fun TEST_clearPrimaryTrafficPercentCache() {
```

Comment:
The original implementation was intentionally written to avoid having test-use-only methods – which I saw as an antipattern/smell.

Can you elaborate on the benefits of this `TEST_` approach over the previous implementation?

In the iOS SDK, we use a `useCache: Boolean = true` to test against the cache-less code path

Reply:
Great point! I mentioned this on standup today but your question hear motivated me to look a bit deeper into why we were even caching `getVaultProvider` in the first place. It turns out we no longer need to do this so I updated the PR to remove all test-specific functionality. Not sure if the same came be done in iOS but glad you brought this up - `LDManager` should be much simpler now!